utils/land-patterns.stanza

```diff
@@ -1040,82 +1040,165 @@ public pcb-landpattern smd-testpoint-pkg (d:Double) :
   layer(Silkscreen("f-silk", Top)) = Circle(d / 2.0)
   ref-label()
 
+; Helper struct to contain the results of the IPC formula in section 3 of 
+; 7351-B
+defstruct IpcResults:
+  Zmax:Double,
+  Gmin:Double,
+  Xmin:Double,
+
+; Helper to compute the pad size from the formula results
+defn pad-size (i:IpcResults) -> Dims:
+  Dims(0.5 * (Zmax(i) - Gmin(i)), Xmin(i))
+
+; the IPC formula
+defn ipc-formula (
+  L:Toleranced, ; the distance from edge-of-lead to edge-of-lead on the exterior of the land pattern
+  T:Toleranced, ; the nominal size of the leads, in the same dimension as L
+  W:Toleranced, ; the nominal size of hte leads, in the dimension orthogonal to L
+  fillets:LeadFillets, ; looked up using lead-fillets(lead-type, density-level)
+):
+  ; compute root-mean-squared of a sequence of numbers
+  val [Jt, Jh, Js, _] = to-tuple(fillets)
+  val Lmax = max-value(L)
+  val Lmin = min-value(L)
+  val Wmin = min-value(W)
+  val Tmin = min-value(T)
+  val Smax = Lmax - 2.0 * Tmin
+  val C_L  = tolerance-range(L)
+  val C_W  = tolerance-range(W)
+  val C_T  = tolerance-range(T)
+  val C_S  = sqrt(C_L * C_L + C_T * C_T)
+  val Zmax = Lmin + 2.0 * Jt + C_L ; the distance from edge of land to edge of land on the exterior of the land pattern
+  val Gmin = Smax - 2.0 * Jh - C_S ; the distance from edge of land to edge of land on the interior of the land pattern
+  val Xmin = Wmin + 2.0 * Js + C_W ; the size of the land in the dimension orthogonal to Z and G.
+  IpcResults(Zmax, Gmin, Xmin)
+   
+; Create an `n` pin SOIC land pattern. 
+;
+; Restrictions:
+; - n must be even
+; - the leads are oriented on the left and right sides of the land pattern
+; - p[1] is the top left pin
+;
+public defn make-n-pin-soic-landpattern (num-pins:Int,       ; number of pins of the component
```

Comment:
as a user, outer-x part-x etc. are not terms I've seen before

Reply:
I agree it's a bad name. I'm probably over-optimizing for clarity. 7351 defines the `outer-x` value as "the overall length of the component" and `part-x` is not used at all (it is only used as to compute the polarity marker location). I'm using `-x` and `-y` to be absolutely clear which dimension the function is using them in, to avoid ambiguity with "length" and "width" (for example, the "length" of a lead/termination is usually orthogonal to the "length" of the component). 

I'd love to use better names as long as we avoid single letter variables, which are even more difficult to parse as a human.

Reply:
IPC-7351 and JEDEC 95 use the same letters for different things - example, L as terminal Length (JEDEC) or land to land length (IPC). I'll share a useful JEDEC doc for definitions of dimensions. We can probably abbreviate those to something folks will quickly recognize.